### PR TITLE
Floating-Points

### PR DESCRIPTION
**Pull request by lukfried**
Added Expr::is_nan, removed BackendFunc and BackendFuncApp, adressed comments of previous PR #567

### CODE REVIEW

File: prusti-common/src/vir/ast/expr.rs [hunk 1]
```diff
@@ -4,6 +4,8 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
+use viper::AstFactory;
```
Comment (vakaras) [inline prusti-common/src/vir/ast/expr.rs]: Why did you add this?

File: prusti-tests/tests/verify/fail/floats/nan-cmp-issues/nan_eq_nan_f64.rs [hunk 2]
```diff
@@ -0,0 +1,5 @@
+// ignore-test: link to the issue
```
Comment (vakaras) [inline prusti-tests/tests/verify/fail/floats/nan-cmp-issues/nan_eq_nan_f64.rs]:
```suggestion
// ignore-test: lhttps://github.com/viperproject/prusti-dev/issues/575
```

File: prusti-tests/tests/verify/fail/floats/nan-cmp-issues/nan_eq_nan_f64.rs [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+fn main() {
+    assert!(f64::NAN == f64::NAN); //~ ERROR: the asserted expression might not hold
+    // assert!(f64::NAN != f64::NAN);
```
Comment (vakaras) [inline prusti-tests/tests/verify/fail/floats/nan-cmp-issues/nan_eq_nan_f64.rs]:
```suggestion
```

File: prusti-tests/tests/verify/fail/floats/nan-cmp-issues/nan_eq_nan_f64.rs (hunk 2 above)
Comment (vakaras) [reply prusti-tests/tests/verify/fail/floats/nan-cmp-issues/nan_eq_nan_f64.rs]: Also, in all other files.

File: prusti-tests/tests/verify/fail/floats/nan_cmp_nan.rs [hunk 4]
```diff
@@ -0,0 +1,40 @@
+// ignore-test: link to the issue
```
Comment (vakaras) [inline prusti-tests/tests/verify/fail/floats/nan_cmp_nan.rs]: Since `f32` works, could you please split this into two files: one for `f32` and one for `f64`?

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 5]
```diff
@@ -2205,7 +2201,12 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
                             // Store a label for permissions got back from the call
                             self.label_after_location.insert(location, label.clone());
 
-                            let lhs = vir::Expr::Field(std::boxed::Box::from(target_place.clone()), vir::Field::new("val_bool", vir::Type::Bool), vir::Position::default());
+                            let lhs = vir::Expr::field(target_place.clone(),vir::Field::new("val_bool", vir::Type::Bool));
+                            // let lhs = vir::Expr::Field(std::boxed::Box::from(target_place.clone()), vir::Field::new("val_bool", vir::Type::Bool), vir::Position::default());
```
Comment (vakaras) [inline prusti-viper/src/encoder/procedure_encoder.rs]:
```suggestion
```
Please do not have commented out lines.

**Review by vakaras — CHANGES_REQUESTED** [review]
I left some comments.

File: prusti-common/src/vir/ast/expr.rs (hunk 1 above)
Comment (lukfried) [reply prusti-common/src/vir/ast/expr.rs]: I think it was auto-completed when I tried things out. I removed it again.

**Comment by vakaras** [comment]
@lukfried Could you please fix the CI failure?

**Review by vakaras — APPROVED** [review]
